### PR TITLE
max timelines in dcs

### PR DESCRIPTION
**Pull request by ksarabu1**
### timeline history consuming space in dcs

This is to address the issue from #[1487](https://github.com/zalando/patroni/issues/1487). A new global parameter max_timelines_history for keeping maximum timelines history in DCS. Default: 0. When set to 0, it keeps all timelines history in dcs.

### CODE REVIEW

**Comment by ksarabu1** [comment]
Looks like its the issue with test case. [Switchover](https://patroni.readthedocs.io/en/latest/pause.html#the-implementation) supposed to go through in paused mode when leader and candidate info is provided.

```
curl -u xyz:xyzxyz -X POST http://pgtest132008/switchover --data '{"leader": "pgtest1"}'
Switchover is possible only to a specific candidate in a paused state
```

File: patroni/postgresql/misc.py [hunk 1]
```diff
@@ -58,8 +58,12 @@ def parse_lsn(lsn):
...
 
 
-def parse_history(data):
-    for line in data.split('\n'):
+def parse_history(data, max_timelines=0):
+    timelines = None
+    if max_timelines is not None and abs(max_timelines) > 0:
+        # twice the requested as the timeline history has empty lines
+        timelines = -abs(max_timelines*2)
+    for line in data.split('\n')[timelines:]:
```
Comment (CyberDem0n) [inline patroni/postgresql/misc.py:66]: 1. I wouldn't rely on empty lines, because there is no guaranty that it will not change in the future.
2. I also think that passing the `max_timelines` argument all the way down to this function is unnecessary. Do we really want to save a few cpu cycles on parsing the complete history file? It would be good enough to just change the https://github.com/zalando/patroni/pull/1491/files#diff-57c53e988b775ee2902488fb33c1f433R525 to
```python
               for line in history[-max_timelines:]:
```

File: patroni/postgresql/__init__.py [hunk 2]
```diff
@@ -801,6 +801,15 @@ def remove_data_directory(self):
...
 
+                pg_tblsp_path = os.path.join(self._data_dir, 'pg_tblspc')
+                if os.path.exists(pg_tblsp_path):
+                    for tspdn in [os.path.join(pg_tblsp_path, dn) for dn in os.listdir(pg_tblsp_path)
+                                  if parse_int(dn) and os.path.islink(os.path.join(pg_tblsp_path, dn))]:
+                        pg_tsp_path = os.path.realpath(tspdn)
+                        if pg_tsp_path is not None and os.path.exists(pg_tsp_path):
+                            logger.info('Removing user defined tablespace directory: %s', pg_tsp_path)
+                            shutil.rmtree(pg_tsp_path, ignore_errors=True)
+
```
Comment (CyberDem0n) [inline patroni/postgresql/__init__.py:812]: This change is irrelevant to max timelines, please move it into a separate PR.
Also using the list comprehension is a bit inappropriate in the for loop, because later you are doing additional checks anyway.

File: patroni/postgresql/misc.py (hunk 1 above)
Comment (ksarabu1) [reply patroni/postgresql/misc.py:66]: 1. Agree. We can move that logic to get_history to trim timelines.
2. Will update.

File: patroni/postgresql/__init__.py (hunk 2 above)
Comment (ksarabu1) [reply patroni/postgresql/__init__.py:812]: Sure. Will submit new PR for this change.

File: patroni/dcs/__init__.py [hunk 3]
```diff
@@ -535,6 +535,10 @@ def timeline(self):
                 return 1
         return 0
 
+    @property
+    def max_timelines_history(self):
+        return self.config.data.get('max_timelines_history', 0)
+
```
Comment (CyberDem0n) [inline patroni/dcs/__init__.py:541]: The `max_timelines_history` is a configuration parameter, so it is more appropriate to put it into the [ClusterConfig](https://github.com/zalando/patroni/pull/1491/files#diff-0964cdd11e7e37271af75805b15fbcd2R319) class.

File: patroni/ha.py [hunk 4]
```diff
@@ -514,12 +514,13 @@ def while_not_sync_standby(self, func):
     def update_cluster_history(self):
         master_timeline = self.state_handler.get_master_timeline()
         cluster_history = self.cluster.history and self.cluster.history.lines
+
```
Comment (CyberDem0n) [inline patroni/ha.py:517]: Unrelated change.

File: patroni/dcs/__init__.py (hunk 3 above)
Comment (ksarabu1) [reply patroni/dcs/__init__.py:541]: Thanks. Updating the code.

File: patroni/ha.py (hunk 4 above)
Comment (ksarabu1) [reply patroni/ha.py]: Thanks. Removed.

File: patroni/postgresql/__init__.py [hunk 5]
```diff
@@ -679,7 +679,7 @@ def replica_cached_timeline(self, master_timeline):
     def get_master_timeline(self):
         return self._cluster_info_state_get('timeline')
 
-    def get_history(self, timeline):
+    def get_history(self, timeline, max_timelines=0):
```
[the remark this answers is not in the document]
Comment (ksarabu1) [reply patroni/postgresql/__init__.py]: Thanks. Removed.

File: docs/SETTINGS.rst [hunk 6]
```diff
@@ -15,6 +15,7 @@ Dynamic configuration is stored in the DCS (Distributed Configuration Store) and
 -  **ttl**: the TTL to acquire the leader lock (in seconds). Think of it as the length of time before initiation of the automatic failover process. Default value: 30
 -  **retry\_timeout**: timeout for DCS and PostgreSQL operation retries (in seconds). DCS or network issues shorter than this will not cause Patroni to demote the leader. Default value: 10
 -  **maximum\_lag\_on\_failover**: the maximum bytes a follower may lag to be able to participate in leader election.
+-  **max\_timelines\_history**: maximum timelines history should be kept in DCS. Default value: 0. When set to 0, it keeps all timelines history in dcs.
```
Comment (avaczi) [inline docs/SETTINGS.rst:18]: I suggest changing the wording to 
'...: maximum number of timeline history items kept in DCS.  Default value: 0. When set to 0, it keeps the full history in DCS.'

**Comment by Jan-M** [comment]
Looks good and also appreciated making this configurable. Difficult to anticipate what other users expect as a good default.

**Comment by Jan-M** [comment]
:+1:

**Comment by CyberDem0n** [comment]
:+1:
